File: ios/RNRadarUtils.m
#import "RNRadarUtils.h"
#import <React/RCTConvert.h>

@implementation RNRadarUtils

+ (NSString *)stringForPermissionsStatus:(CLAuthorizationStatus)status {
    switch (status) {
        case kCLAuthorizationStatusDenied:
            return @"DENIED";
        case kCLAuthorizationStatusRestricted:
            return @"DENIED";
        case kCLAuthorizationStatusAuthorizedAlways:
            return @"GRANTED";
        case kCLAuthorizationStatusAuthorizedWhenInUse:
            return @"GRANTED";
        default:
            return @"UNKNOWN";
    }
}

+ (NSString *)stringForStatus:(RadarStatus)status {
    switch (status) {
        case RadarStatusSuccess:
            return @"SUCCESS";
        case RadarStatusErrorPublishableKey:
            return @"ERROR_PUBLISHABLE_KEY";
        case RadarStatusErrorPermissions:
            return @"ERROR_PERMISSIONS";
        case RadarStatusErrorLocation:
            return @"ERROR_LOCATION";
        case RadarStatusErrorNetwork:
            return @"ERROR_NETWORK";
        case RadarStatusErrorUnauthorized:
            return @"ERROR_UNAUTHORIZED";
        case RadarStatusErrorRateLimit:
            return @"ERROR_RATE_LIMIT";
        case RadarStatusErrorServer:
            return @"ERROR_SERVER";
        default:
            return @"ERROR_UNKNOWN";
    }
}

+ (NSString *)stringForEventType:(RadarEventType)type {
    switch (type) {
        case RadarEventTypeUserEnteredGeofence:
            return @"user.entered_geofence";
        case RadarEventTypeUserExitedGeofence:
            return @"user.exited_geofence";
        case RadarEventTypeUserEnteredHome:
            return @"user.entered_home";
        case RadarEventTypeUserExitedHome:
            return @"user.exited_home";
        case RadarEventTypeUserEnteredOffice:
            return @"user.entered_office";
        case RadarEventTypeUserExitedOffice:
            return @"user.exited_office";
        case RadarEventTypeUserStartedTraveling:
            return @"user.started_traveling";
        case RadarEventTypeUserStoppedTraveling:
            return @"user.stopped_traveling";
        case RadarEventTypeUserEnteredPlace:
            return @"user.entered_place";
        case RadarEventTypeUserExitedPlace:
            return @"user.exited_place";
        case RadarEventTypeUserNearbyPlaceChain:
            return @"user.nearby_place_chain";
        default:
            return nil;

    }
}

+ (NSNumber *)numberForEventConfidence:(RadarEventConfidence)confidence {
    switch (confidence) {
        case RadarEventConfidenceHigh:
            return @(3);
        case RadarEventConfidenceMedium:
            return @(2);
        case RadarEventConfidenceLow:
            return @(1);
        default:
            return @(0);
    }
}

+ (NSString *)stringForUserInsightsLocationType:(RadarUserInsightsLocationType)type {
    switch (type) {
        case RadarUserInsightsLocationTypeHome:
            return @"home";
        case RadarUserInsightsLocationTypeOffice:
            return @"office";
        default:
            return nil;
    }
}

+ (NSNumber *)numberForUserInsightsLocationConfidence:(RadarUserInsightsLocationConfidence)confidence {
    switch (confidence) {
        case RadarUserInsightsLocationConfidenceHigh:
            return @(3);
        case RadarUserInsightsLocationConfidenceMedium:
            return @(2);
        case RadarUserInsightsLocationConfidenceLow:
            return @(1);
        default:
            return @(0);
    }
}

+ (RadarPlacesProvider)placesProviderForString:(NSString *)providerStr {
    if ([providerStr isEqualToString:@"facebook"]) {
        return RadarPlacesProviderFacebook;
    }
    return RadarPlacesProviderNone;
}

+ (NSDictionary *)dictionaryForUser:(RadarUser *)user {
    if (!user) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    [dict setValue:user._id forKey:@"_id"];
    [dict setValue:user.userId forKey:@"userId"];
    NSString *description = user._description;
    if (description) {
        [dict setValue:description forKey:@"description"];
    }
    NSMutableArray *geofencesArr = [[NSMutableArray alloc] initWithCapacity:user.geofences.count];
    for (RadarGeofence *geofence in user.geofences) {
        NSDictionary *geofenceDict = [self dictionaryForGeofence:geofence];
        [geofencesArr addObject:geofenceDict];
    }
    [dict setValue:geofencesArr forKey:@"geofences"];
    NSDictionary *insightsDict = [self dictionaryForUserInsights:user.insights];
    [dict setValue:insightsDict forKey:@"insights"];
    if (user.place) {
      NSDictionary *placeDict = [self dictionaryForPlace:user.place];
      [dict setValue:placeDict forKey:@"place"];
    }
    return dict;
}

+ (NSDictionary *)dictionaryForUserInsights:(RadarUserInsights *)insights {
    if (!insights) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    NSDictionary *homeLocationDict = [self dictionaryForUserInsightsLocation:insights.homeLocation];
    if (homeLocationDict) {
        [dict setObject:homeLocationDict forKey:@"homeLocation"];
    }
    NSDictionary *officeLocationDict = [RNRadarUtils dictionaryForUserInsightsLocation:insights.officeLocation];
    if (officeLocationDict) {
        [dict setObject:officeLocationDict forKey:@"officeLocation"];
    }
    NSDictionary *stateDict = [self dictionaryForUserInsightsState:insights.state];
    if (stateDict) {
        [dict setObject:stateDict forKey:@"state"];
    }
    return dict;
}

+ (NSDictionary *)dictionaryForUserInsightsLocation:(RadarUserInsightsLocation *)location {
    if (!location) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    NSString *type = [RNRadarUtils stringForUserInsightsLocationType:location.type];
    if (type) {
        [dict setValue:type forKey:@"type"];
    }
    NSDictionary *locationDict = [RNRadarUtils dictionaryForLocation:location.location];
    if (locationDict) {
        [dict setValue:locationDict forKey:@"location"];
    }
    NSNumber *confidence = [RNRadarUtils numberForUserInsightsLocationConfidence:location.confidence];
    [dict setValue:confidence forKey:@"confidence"];
    return dict;
}

+ (NSDictionary *)dictionaryForUserInsightsState:(RadarUserInsightsState *)state {
    if (!state) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    [dict setValue:@(state.home) forKey:@"home"];
    [dict setValue:@(state.office) forKey:@"office"];
    [dict setValue:@(state.traveling) forKey:@"traveling"];
    return dict;
}

+ (NSDictionary *)dictionaryForGeofence:(RadarGeofence *)geofence {
    if (!geofence) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    [dict setValue:geofence._id forKey:@"_id"];
    NSString *tag = geofence.tag;
    if (tag) {
        [dict setValue:tag forKey:@"tag"];
    }
    NSString *externalId = geofence.externalId;
    if (externalId) {
        [dict setValue:externalId forKey:@"externalId"];
    }
    [dict setValue:geofence._description forKey:@"description"];
    NSDictionary *metadata = geofence.metadata;
    if (metadata) {
        [dict setValue:metadata forKey:@"metadata"];
    }
    return dict;
}

+ (NSDictionary *)dictionaryForPlace:(RadarPlace *)place {
    if (!place) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    [dict setValue:place._id forKey:@"_id"];
    NSString *facebookId = place.facebookId;
    if (facebookId) {
        [dict setValue:facebookId forKey:@"facebookId"];
    }
    [dict setValue:place.name forKey:@"name"];
    if (place.categories && place.categories.count) {
        NSMutableArray *categories = [[NSMutableArray alloc] initWithCapacity:place.categories.count];
        for (NSString *category in place.categories) {
            [categories addObject:category];
        }
        [dict setValue:categories forKey:@"categories"];
    }
    if (place.chain) {
      NSDictionary *chain = @{@"slug": place.chain.slug, @"name": place.chain.name};
      [dict setValue:chain forKey:@"chain"];
    }
    return dict;
}

+ (NSArray *)arrayForEvents:(NSArray<RadarEvent *> *)events {
    if (!events) {
        return nil;
    }

    NSMutableArray *arr = [[NSMutableArray alloc] initWithCapacity:events.count];
    for (RadarEvent *event in events) {
        NSDictionary *dict = [self dictionaryForEvent:event];
        [arr addObject:dict];
    }
    return arr;
}

+ (NSDictionary *)dictionaryForEvent:(RadarEvent *)event {
    if (!event) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    [dict setValue:event._id forKey:@"_id"];
    [dict setValue:@(event.live) forKey:@"live"];
    NSString *type = [RNRadarUtils stringForEventType:event.type];
    if (type) {
        [dict setValue:type forKey:@"type"];
    }
    NSDictionary *geofenceDict = [self dictionaryForGeofence:event.geofence];
    if (geofenceDict) {
        [dict setValue:geofenceDict forKey:@"geofence"];
    }
    NSDictionary *placeDict = [RNRadarUtils dictionaryForPlace:event.place];
    if (placeDict) {
        [dict setValue:placeDict forKey:@"place"];
    }
    NSNumber *confidence = [RNRadarUtils numberForEventConfidence:event.confidence];
    [dict setValue:confidence forKey:@"confidence"];
    if (event.duration) {
        [dict setValue:@(event.duration) forKey:@"duration"];
    }
    NSArray *alternatePlaces = [self arrayForAlternatePlaces:event.alternatePlaces];
    if (alternatePlaces) {
        [dict setValue:alternatePlaces forKey:@"alternatePlaces"];
    }
    return dict;
}

+ (NSArray *)arrayForAlternatePlaces:(NSArray<RadarPlace *> *)places {
    if (!places) {
        return nil;
    }

    NSMutableArray *arr = [[NSMutableArray alloc] initWithCapacity:places.count];
    for (RadarPlace *place in places) {
        NSDictionary *dict = [RNRadarUtils dictionaryForPlace:place];
        [arr addObject:dict];
    }
    return arr;
}

+ (NSDictionary *)dictionaryForLocation:(CLLocation *)location {
    if (!location) {
        return nil;
    }

    NSMutableDictionary *dict = [NSMutableDictionary new];
    [dict setValue:@(location.coordinate.latitude) forKey:@"latitude"];
    [dict setValue:@(location.coordinate.longitude) forKey:@"longitude"];
    if (location.horizontalAccuracy) {
        [dict setValue:@(location.horizontalAccuracy) forKey:@"accuracy"];
    }
    return dict;
}

+ (RadarTrackingOptions *)optionsForDictionary:(NSDictionary *)optionsDict {
    RadarTrackingOptions *options = [RadarTrackingOptions new];
    if (optionsDict) {
        if (optionsDict[@"sync"]) {
            NSString *sync = [RCTConvert NSString:optionsDict[@"sync"]];
            if ([sync isEqualToString:@"possibleStateChanges"]) {
                options.sync = RadarTrackingSyncPossibleStateChanges;
            } else if ([sync isEqualToString:@"all"]) {
                options.sync = RadarTrackingSyncAll;
            }
        }
        if (optionsDict[@"offline"]) {
            NSString *offline = [RCTConvert NSString:optionsDict[@"offline"]];
            if ([offline isEqualToString:@"replayStopped"]) {
                options.offline = RadarTrackingOfflineReplayStopped;
            } else if ([offline isEqualToString:@"replayOff"]) {
                options.offline = RadarTrackingOfflineReplayOff;
            }
        }
        if (optionsDict[@"priority"]) {
            NSString *priority = [RCTConvert NSString:optionsDict[@"priority"]];
            if ([priority isEqualToString:@"efficiency"]) {
                options.priority = RadarTrackingPriorityEfficiency;
            } else if ([priority isEqualToString:@"responsiveness"]) {
                options.priority = RadarTrackingPriorityResponsiveness;
            }
        }
    }
    return options;
}


@end

File: android/src/main/java/com/onradar/react/RNRadarUtils.java
package com.onradar.react;

import android.location.Location;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.onradar.sdk.Radar;
import com.onradar.sdk.model.RadarEvent;
import com.onradar.sdk.model.RadarGeofence;
import com.onradar.sdk.model.RadarPlace;
import com.onradar.sdk.model.RadarUser;
import com.onradar.sdk.model.RadarUserInsights;
import com.onradar.sdk.model.RadarUserInsightsLocation;
import com.onradar.sdk.model.RadarUserInsightsState;

class RNRadarUtils {

    static String stringForPermissionsStatus(boolean hasGrantedPermissions) {
        if (hasGrantedPermissions)
            return "GRANTED";
        return "DENIED";
    }

    static String stringForStatus(Radar.RadarStatus status) {
        return status.toString();
    }

    static String stringForEventType(RadarEvent.RadarEventType type) {
        switch (type) {
            case USER_ENTERED_GEOFENCE:
                return "user.entered_geofence";
            case USER_EXITED_GEOFENCE:
                return "user.exited_geofence";
            case USER_ENTERED_HOME:
                return "user.entered_home";
            case USER_EXITED_HOME:
                return "user.exited_home";
            case USER_ENTERED_OFFICE:
                return "user.entered_office";
            case USER_EXITED_OFFICE:
                return "user.exited_office";
            case USER_STARTED_TRAVELING:
                return "user.started_traveling";
            case USER_STOPPED_TRAVELING:
                return "user.stopped_traveling";
            case USER_ENTERED_PLACE:
                return "user.entered_place";
            case USER_EXITED_PLACE:
                return "user.exited_place";
            default:
                return null;
        }
    }

    static int numberForEventConfidence(RadarEvent.RadarEventConfidence confidence) {
        switch (confidence) {
            case HIGH:
                return 3;
            case MEDIUM:
                return 2;
            case LOW:
                return 1;
            default:
                return 0;
        }
    }

    static String stringForUserInsightsLocationType(RadarUserInsightsLocation.RadarUserInsightsLocationType type) {
        switch (type) {
            case HOME:
                return "home";
            case OFFICE:
                return "office";
            default:
                return null;
        }
    }

    static int numberForUserInsightsLocationConfidence(RadarUserInsightsLocation.RadarUserInsightsLocationConfidence confidence) {
        switch (confidence) {
            case HIGH:
                return 3;
            case MEDIUM:
                return 2;
            case LOW:
                return 1;
            default:
                return 0;
        }
    }

    static Radar.RadarPlacesProvider placesProviderForString(String providerStr) {
        if (providerStr.equals("facebook"))
            return Radar.RadarPlacesProvider.FACEBOOK;
        return Radar.RadarPlacesProvider.NONE;
    }

    static WritableMap mapForUser(RadarUser user) {
        if (user == null)
            return null;

        WritableMap map = Arguments.createMap();
        map.putString("_id", user.getId());
        map.putString("userId", user.getUserId());
        String description = user.getDescription();
        if (description != null)
            map.putString("description", description);
        WritableArray geofencesArr = Arguments.createArray();
        for (RadarGeofence geofence : user.getGeofences()) {
            WritableMap geofenceMap = RNRadarUtils.mapForGeofence(geofence);
            geofencesArr.pushMap(geofenceMap);
        }
        map.putArray("geofences", geofencesArr);
        WritableMap insightsMap = RNRadarUtils.mapForUserInsights(user.getInsights());
        if (insightsMap != null)
            map.putMap("insights", insightsMap);
        return map;
    }

    private static WritableMap mapForUserInsights(RadarUserInsights insights) {
        if (insights == null)
            return null;

        WritableMap map = Arguments.createMap();
        WritableMap homeLocationMap = RNRadarUtils.mapForUserInsightsLocation(insights.getHomeLocation());
        if (homeLocationMap != null)
            map.putMap("homeLocation", homeLocationMap);
        WritableMap officeLocationMap = RNRadarUtils.mapForUserInsightsLocation(insights.getOfficeLocation());
        if (officeLocationMap != null)
            map.putMap("officeLocation", officeLocationMap);
        WritableMap stateMap = RNRadarUtils.mapForUserInsightsState(insights.getState());
        if (stateMap != null)
            map.putMap("state", stateMap);
        return map;
    }

    private static WritableMap mapForUserInsightsLocation(RadarUserInsightsLocation location) {
        if (location == null)
            return null;

        WritableMap map = Arguments.createMap();
        String type = RNRadarUtils.stringForUserInsightsLocationType(location.getType());
        if (type != null)
            map.putString("type", type);
        WritableMap locationMap = RNRadarUtils.mapForLocation(location.getLocation());
        if (locationMap != null)
            map.putMap("location", locationMap);
        int confidence = RNRadarUtils.numberForUserInsightsLocationConfidence(location.getConfidence());
        map.putInt("confidence", confidence);
        return map;
    }

    private static WritableMap mapForUserInsightsState(RadarUserInsightsState state) {
        if (state == null)
            return null;

        WritableMap map = Arguments.createMap();
        map.putBoolean("home", state.getHome());
        map.putBoolean("office", state.getOffice());
        map.putBoolean("traveling", state.getTraveling());
        return map;
    }

    private static WritableMap mapForGeofence(RadarGeofence geofence) {
        if (geofence == null)
            return null;

        WritableMap map = Arguments.createMap();
        map.putString("_id", geofence.getId());
        String tag = geofence.getTag();
        if (tag != null)
            map.putString("tag", tag);
        String externalId = geofence.getExternalId();
        if (externalId != null)
            map.putString("externalId", externalId);
        map.putString("description", geofence.getDescription());
        return map;
    }

    private static WritableMap mapForPlace(RadarPlace place) {
        if (place == null)
            return null;

        WritableMap map = Arguments.createMap();
        map.putString("_id", place.getId());
        String facebookId = place.getFacebookId();
        if (facebookId != null)
            map.putString("facebookId", facebookId);
        map.putString("name", place.getName());
        if (place.getCategories() != null && place.getCategories().length > 0) {
            WritableArray categories = Arguments.createArray();
            for (String category : place.getCategories()) {
                categories.pushString(category);
            }
            map.putArray("categories", categories);
        }
        if (place.getChain() != null) {
            WritableMap chain = Arguments.createMap();
            chain.putString("slug", place.getChain().getSlug());
            chain.putString("name", place.getChain().getName());
            map.putMap("chain", chain);
        }
        return map;
    }

    static WritableArray arrayForEvents(RadarEvent[] events) {
        if (events == null)
            return null;

        WritableArray arr = Arguments.createArray();
        for (RadarEvent event : events) {
            WritableMap map = RNRadarUtils.mapForEvent(event);
            arr.pushMap(map);
        }
        return arr;
    }

    private static WritableMap mapForEvent(RadarEvent event) {
        if (event == null)
            return null;

        WritableMap map = Arguments.createMap();
        map.putString("_id", event.getId());
        map.putBoolean("live", event.getLive());
        String type = RNRadarUtils.stringForEventType(event.getType());
        if (type != null)
            map.putString("type", type);
        WritableMap geofenceMap = RNRadarUtils.mapForGeofence(event.getGeofence());
        if (geofenceMap != null)
            map.putMap("geofence", geofenceMap);
        WritableMap placeMap = RNRadarUtils.mapForPlace(event.getPlace());
        if (placeMap != null)
            map.putMap("place", placeMap);
        int confidence = RNRadarUtils.numberForEventConfidence(event.getConfidence());
        map.putInt("confidence", confidence);
        if (event.getDuration() != 0)
            map.putDouble("duration", event.getDuration());
        return map;
    }

    static WritableMap mapForLocation(Location location) {
        if (location == null)
            return null;

        WritableMap map = Arguments.createMap();
        map.putDouble("latitude", location.getLatitude());
        map.putDouble("longitude", location.getLongitude());
        if (location.getAccuracy() != 0)
            map.putDouble("accuracy", location.getAccuracy());
        return map;
    }

}
